### PR TITLE
Agones Unity SDK development setup instructions + Agones Unity SDK Ready test

### PR DESCRIPTION
**Pull request by aallbrig**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation
/kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:
Following instructions proposed in https://github.com/googleforgames/agones/pull/3886 I set up a local environment to contribute to agones unity client sdk.

I noticed there wasn't a test. This adds a first test. I don't think it's a good test so please offer up what might make for a better first test.

**Which issue(s) this PR fixes**:

Closes #

**Special notes for your reviewer**:

### CODE REVIEW

File: sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    public class AgonesSdkComplianceTests
+    {
+        [UnityTest]
+        public IEnumerator AgonesSdk_Ready_AlwaysReturnsTrue()
+        {
+            var gameObject = new GameObject();
+            var sut = gameObject.AddComponent<AgonesSdk>();
+            var task = sut.Ready();
+            yield return AwaitTask(task);
+            Assert.IsTrue(task.Result);
```
Comment (markmandel) [inline sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs:19]: This is not a bad test (although we may get double coverage if we ever do a conformance test - which is not well documented)  - but wondering if this passes, as there is no mock for the http endpoint for the `/ready` request?

**Comment by igooch** [comment]
Noob question -- what are the steps to run the tests? Are we able to run the tests similar to vanilla csharp (csharp SDK tests have the non-obvious steps of 1. navigating to `agones/build` 2. running `make sdk-shell-csharp` 3. once in the C# shell navigating to the test folder /go/src/agones.dev/agones/sdks/csharp/test 4. using standard dotnet commands such as `dotnet test`.), or does this require a Unity editor?

**Comment by ZeroParticle** [comment]
> Noob question -- what are the steps to run the tests? Are we able to run the tests similar to vanilla csharp (csharp SDK tests have the non-obvious steps of 1. navigating to `agones/build` 2. running `make sdk-shell-csharp` 3. once in the C# shell navigating to the test folder /go/src/agones.dev/agones/sdks/csharp/test 4. using standard dotnet commands such as `dotnet test`.), or does this require a Unity editor?

For the tests setup I added in #3883, I added a README that might help answer this for you. It's at the very bottom of the PR. My understanding is that it's basically impossible to run these tests without the unity editor. Even the common CI setups for Unity are just running a dockerized editor.

**Comment by igooch** [comment]
> My understanding is that it's basically impossible to run these tests without the unity editor. Even the common CI setups for Unity are just running a dockerized editor.

Would it makes sense to open up a new issue for creating some sort of pipeline for running the Unity tests? Possibly CloudBuild could take in a Unity license as a Secret and run the tests in the dockerized editor mode?

**Comment by aallbrig** [comment]
@igooch You're definitely right that developing unity packages is not intuitive. I wrote documentation in this PR (https://github.com/googleforgames/agones/pull/3886) for how to set up a testing environment that's common. I've used this setup when contributing to other unity open source projects like Mirror. You can follow the instructions in that PR to set up to then develop a unity package.

Should I merge that PR into this one, so documentation exists along with this test?

**Comment by aallbrig** [comment]
@igooch I can help with using game CI's docker container to run tests. I've used that container image before to cross platform test & build in parallel since I didn't like serially builds in unity editor.

There's a little setup to do but I've scripted its use out before. Would need to review this project's CI more to understand how to introduce a new testing step (guidance is welcome and appreciated!)

Unity docker usage seems outside the scope of this PR, maybe, since it's a test suite agnostic enhancement? WDYT?

File: sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            _mockSdkServer.RegisterResponseHandler("/ready", _ => "{}");
+            var sut = _gameObject.AddComponent<AgonesSdk>();
+            var task = sut.Ready();
+            yield return AwaitTask(task);
+            Assert.IsTrue(task.Result);
+            _mockSdkServer.DeregisterResponseHandler("/ready");
+        }
+        [UnityTest]
+        public IEnumerator AgonesSdk_()
```
[the remark this answers is not in the document]
Comment (aallbrig) [reply sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs]: Done :)

**Comment by aallbrig** [comment]
Merged in the contents of https://github.com/googleforgames/agones/pull/3886 into this PR, to provide instructions for this test.

In order to simplify the setup instructions I created a new Unity project in `test/sdk/unity`, which is very much inspired by @ZeroParticle's https://github.com/googleforgames/agones/pull/3883. The difference between this implementation and @ZeroParticle's is that I keep the test co-located with the Agones Unity SDK custom package which this new test unity project pulls in (relative path in `Packages/manifest.json`), whereas @ZeroParticle has their test in the Unity project which isn't standard to typical Unity Custom Packages (See [Unity Custom Package Layout](https://docs.unity3d.com/Manual/cus-layout.html)).

The PR is still WIP; now that I can have some guarantees with the Unity Project I can install some additional testing libraries more easily. Instead of having `sdks/unity/Tests/TestingEnvironment/MockAgonesSdkServer.cs` I could use a 3rd party library to stub out `UnityWebRequest` for tests.

**Comment by aallbrig** [comment]
@igooch @markmandel et all This PR is ready to be reviewed!
Features:

- New Unity project created at `test/sdk/unity` that's already _pre-configured_ to pull in local Agones Unity SDK locally (see `test/sdk/unity/Package/manifest.json` so see how). This serves as the host Unity project, which is required for Unity Custom Package development.
- Instructions added on how to develop the Agones Unity SDK added so that developers know how to do things like add the `test/sdk/unity` project to their Unity Hub, and how development will work.
- A single test that spies to ensure `AgonesSdk` component makes request to the Agones SDK Server. I want to add more tests but hope for feedback. View `sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs` to view the test!


Screenshot 2024-07-20 at 1 42 41 PM

The PR is well over 1000 LoC primarily due to having a new Unity project contained in the changeset. See below to see that a majority of this PR's LoC come from default files created when initializing a new Unity project.

```
$ git diff --numstat main...agones-unity-client-sdk-initialize-test | sort -k1,1nr

671     0       test/sdk/unity/ProjectSettings/ProjectSettings.asset
322     0       test/sdk/unity/Packages/packages-lock.json
322     0       test/sdk/unity/ProjectSettings/QualitySettings.asset
316     0       test/sdk/unity/Assets/Main.unity
296     0       test/sdk/unity/ProjectSettings/InputManager.asset
121     0       test/sdk/unity/ProjectSettings/SceneTemplateSettings.json
93      0       test/sdk/unity/ProjectSettings/NavMeshAreas.asset
72      0       test/sdk/unity/.gitignore
67      0       test/sdk/unity/ProjectSettings/GraphicsSettings.asset
50      0       sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs
48      0       test/sdk/unity/ProjectSettings/Physics2DSettings.asset
47      0       test/sdk/unity/ProjectSettings/EditorSettings.asset
44      0       test/sdk/unity/Packages/manifest.json
43      0       test/sdk/unity/ProjectSettings/TagManager.asset
40      0       test/sdk/unity/ProjectSettings/DynamicsManager.asset
36      0       test/sdk/unity/ProjectSettings/PackageManagerSettings.asset
36      0       test/sdk/unity/ProjectSettings/UnityConnectSettings.asset
35      0       test/sdk/unity/ProjectSettings/MemorySettings.asset
25      0       sdks/unity/Tests/TestingEnvironment/SpyRequestSender.cs
23      0       sdks/unity/Tests/Runtime/PlayMode/Tests.Runtime.Agones.asmdef
20      0       test/sdk/unity/ProjectSettings/AudioManager.asset
18      0       test/sdk/unity/ProjectSettings/VFXManager.asset
17      12      sdks/unity/AgonesSdk.cs
14      0       sdks/README.md
14      0       sdks/unity/README.md
12      2       sdks/unity/Agones.asmdef
10      0       sdks/unity/IRequestSender.cs
9       0       test/sdk/unity/ProjectSettings/TimeManager.asset
8       0       sdks/unity/Tests/Runtime/PlayMode.meta
8       0       test/sdk/unity/ProjectSettings/EditorBuildSettings.asset
8       0       test/sdk/unity/ProjectSettings/VersionControlSettings.asset
7       0       sdks/unity/README.md.meta
7       0       test/sdk/unity/Assets/Main.unity.meta
7       0       test/sdk/unity/ProjectSettings/PresetManager.asset
6       0       test/sdk/unity/ProjectSettings/ClusterInputManager.asset
5       0       test/sdk/unity/README.md
3       0       sdks/unity/IRequestSender.cs.meta
3       0       sdks/unity/Tests.meta
3       0       sdks/unity/Tests/Runtime.meta
3       0       sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs.meta
3       0       sdks/unity/Tests/Runtime/PlayMode/Tests.Runtime.Agones.asmdef.meta
3       0       sdks/unity/Tests/TestingEnvironment.meta
3       0       sdks/unity/Tests/TestingEnvironment/SpyRequestSender.cs.meta
2       0       README.md
2       0       test/sdk/unity/ProjectSettings/ProjectVersion.txt
```
Without the `.asset`, `.meta`, `package-lock.json`, and `.unity` files this change set would probably be considered reasonable.
```
$ git diff --numstat main...agones-unity-client-sdk-initialize-test | grep -v '!*.asset\|!*.meta\|!*-lock.json\|!*\.unity' | sort -k1,1nr

121     0       test/sdk/unity/ProjectSettings/SceneTemplateSettings.json
72      0       test/sdk/unity/.gitignore
50      0       sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs
44      0       test/sdk/unity/Packages/manifest.json
25      0       sdks/unity/Tests/TestingEnvironment/SpyRequestSender.cs
23      0       sdks/unity/Tests/Runtime/PlayMode/Tests.Runtime.Agones.asmdef
17      12      sdks/unity/AgonesSdk.cs
14      0       sdks/README.md
14      0       sdks/unity/README.md
12      2       sdks/unity/Agones.asmdef
10      0       sdks/unity/IRequestSender.cs
5       0       test/sdk/unity/README.md
2       0       README.md
2       0       test/sdk/unity/ProjectSettings/ProjectVersion.txt
```

**Comment by aallbrig** [comment]
@zmerlynn @igooch @markmandel tagging for some review. Thank you in advance!

File: sdks/unity/IRequestSender.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+using System.Threading.Tasks;
```
Comment (markmandel) [inline sdks/unity/IRequestSender.cs:1]: Apache licence header please (on everything that's code).

File: sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
+// Copyright 2022 Google LLC
```
Comment (markmandel) [inline sdks/unity/Tests/Runtime/PlayMode/AgonesSdkComplianceTests.cs:1]:
```suggestion
// Copyright 2024 Google LLC
```

File: test/sdk/unity/.gitignore [hunk 5]
```diff
@@ -0,0 +1,72 @@
+# This .gitignore file should be placed at the root of your Unity project directory
```
Comment (markmandel) [inline test/sdk/unity/.gitignore:1]: Apache licence please.

File: sdks/unity/Tests/TestingEnvironment/SpyRequestSender.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+using System.Threading.Tasks;
```
Comment (markmandel) [inline sdks/unity/Tests/TestingEnvironment/SpyRequestSender.cs:1]: Apache licence please.

File: sdks/unity/IRequestSender.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+using UnityEngine.Networking;
+
+namespace Agones
+{
+    public interface IRequestSender
```
Comment (markmandel) [inline sdks/unity/IRequestSender.cs:6]: oooh, what's this do? 😄 👀

File: sdks/unity/README.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Agones Unity Client SDK
```
Comment (markmandel) [inline sdks/unity/README.md:1]: 👍🏻  Noice.

**Review by markmandel — COMMENTED** [review]
Thanks for doing this! Mostly nit comments.

File: sdks/unity/IRequestSender.cs (hunk 3 above)
Comment (aallbrig) [reply sdks/unity/IRequestSender.cs:16]: Whoops! I think I need to see if I can configure my editor to automatically add these headers.

**Comment by aallbrig** [comment]
@markmandel I added an apache header to all added C# code in this PR. I also added it to `sdks/unity/model/StatusAddresses.cs` which is a class that pre-exists this PR but I found it was missing an apache header.

I debated if I should mark the year 2019 or 2024; I opted for 2024.

**Comment by ZeroParticle** [comment]
Happy to update my PR #3883 to use this project if that's what we go for. The tests in my project run against a local SDK server rather than a mock. That seems complimentary to what you have here, so I figure we will want to keep it around rather than replace it.

**Comment by aallbrig** [comment]
@ZeroParticle Thanks for weighing in! Yes I think adding tests in the style of this PR is better than in the style of your PR. The style of this PR has the tests contained within the package files which is more expected for a custom unity package.

I'm hoping @markmandel (and/or other(s)) review/merge this PR soon which will enable both of us to evolve this base code.

**Comment by zmerlynn** [comment]
@ZeroParticle @aallbrig - @markmandel and I were chatting. We don't have a great way to test this right now, unfortunately. @ZeroParticle, if I review this, would you be up for ad hoc testing and seeing if it looks reasonable? @aallbrig - same question for other Unity PRs as they come in.

Basically, if y'all are willing, we're happy to review and then gradually let y'all be the Unity maintainers, but we may need help on the testing front.

**Comment by aallbrig** [comment]
Was trying to get gameci docker container working but got stuck. They changed how licenses work in Unity, and a license is required to be able to use the Unity editor app. Gameci is based on Ubuntu; a hub generated license is tied to the hardware/OS -- so using my M2 license won't work for an Ubuntu container.

Unity changed the process of obtaining a license. It's generated through Unity Hub app now.

I was reverse engineering what gameci does. There's plenty of JS that obfuscates so I just need to keep reading and experimenting. There is a way of generating docker commands to run tests (I've done it before), just need to work out the license limitations.

The reason I bring up gameci is because it could help make running tests something that can run in a container, which may decrease the burden of contributing to Agones Unity SDK. Imagine recommending a user run a few commands instead of downloading/installing Unity Hub, mucking with a GUI, then opening the project in Unity Editor to then run some tests.

Just posting some updates.

I'm still hoping that this PR is deemed mergeable even without the gameci docker commands. Getting the initial Unity project set up and configured to pull in the local copy of Agones Unity SDK is foundational to further development. Even though it would be better to have shareable docker commands, we still have a way of developing/testing the Agones Unity SDK as documented in this PR.

**Review by zmerlynn — APPROVED** [review]

**Comment by zmerlynn** [comment]
I'm happy to take this as a base PR to build on. Thanks @aallbrig and @ZeroParticle !

**Comment by aallbrig** [comment]
Woohoo!

Cc: @ZeroParticle it's been merged! Let me know if you need/want any help on your Agones Unity SDK additions/tests.

**Pull request merged**